lib/SPIRVProducerPass.cpp

```diff
@@ -4434,6 +4461,159 @@ void SPIRVProducerPassImpl::GenerateInstruction(Instruction &I) {
         Ops.clear();
         Ops << Ty << Cmp << Mod << Modm1;
         RID = addSPIRVInst(spv::OpSelect, Ops);
+      } else if (I.getOpcode() == Instruction::FDiv && !Option::UnsafeMath() &&
+                 I.getType()->getScalarType()->isFloatTy()) {
+        // float div(float a, float b) {
+        //   if (abs(b) > 0x1.0p+126f) {
+        //     c = 0x1.0p-4f;
+        //   } else if (abs(b) < 0x1.0p-126f) {
+        //     c = 0x1.0p+24f;
+        //   } else {
+        //     c = 1.0;
+        //   }
+        //   return (a / (b * c)) * c;
+        // }
+        Type *FPTy = I.getType();
+        auto a = getSPIRVValue(I.getOperand(0));
+        auto b = getSPIRVValue(I.getOperand(1));
+        Type *boolTy = IntegerType::get(module->getContext(), 1);
+        if (FPTy->isVectorTy()) {
+          boolTy = VectorType::get(boolTy, dyn_cast<VectorType>(FPTy));
+        }
+        SPIRVOperandVec Ops;
+        SPIRVID Ty = getSPIRVType(FPTy);
+
+
+        const float fmax_val = 0x1.0p+126f;
+        const float fmin_val = 0x1.0p-126f;
+        const float c_max_val = 0x1.0p-4f;
+        const float c_min_val = 0x1.0p+24f;
+        SPIRVID divisor, c;
+
+        // Try to optimize it if b is constant.
+        if (auto b_cst = dyn_cast<ConstantFP>(I.getOperand(1))) {
```

Comment:
There's a lot of repeated code here. What about something along the lines of:
```
const bool is_vector = FPTy->isVectorTy();
const auto num_eles = is_vector ? cast<VectorType>(FPTy)->getNumElements() : 1;
if (auto *cst = dyn_cast<Constant>(I.getOperand(1))) {
  SmallVector<Constant *, 2> constants;
  SmallVector<Constant *, 2> divisor_vec;
  SmallVector<Constant *, 2> c_vec;
  if (is_vector) {
    for (int i = 0; i < num_eles; ++i) {
      constants.push_back(cst->getAggregateElement(i));
    }
  } else {
    constants.push_back(cst);
  }

  bool c_vec_needed = false;
  for (auto *ele : constants) {
    // Process like vector case
  }

  // Construct final values
}
```

If this is less code it's probably worthwhile for readability.

Reply:
indeed, it works well, thanks